src/methods/iterator.js

```diff
@@ -98,6 +107,89 @@ export function IteratorNext(realm: Realm, iterator: Value, value?: Value): Obje
   return result;
 }
 
+// ECMA262 7.4.8
+export function CreateListIterator(realm: Realm, list: Array<Value>): ObjectValue {
+  // 1. Let iterator be ObjectCreate(%IteratorPrototype%, « [[IteratorNext]], [[IteratedList]], [[ListIteratorNextIndex]] »).
+  let iterator = ObjectCreate(realm, realm.intrinsics.IteratorPrototype, {
+    $IteratorNext: undefined,
+    $IteratedList: undefined,
+    $ListIteratorNextIndex: undefined,
+  });
+
+  // 2. Set iterator's [[IteratedList]] internal slot to list.
+  iterator.$IteratedList = list;
+
+  // 3. Set iterator's [[ListIteratorNextIndex]] internal slot to 0.
+  iterator.$ListIteratorNextIndex = 0;
+
+  // 4. Let next be a new built-in function object as defined in ListIterator next (7.4.8.1).
+  let next = ListIterator_next(realm);
+
+  // 5. Set iterator's [[IteratorNext]] internal slot to next.
+  iterator.$IteratorNext = next;
+
+  // 6. Perform CreateMethodProperty(iterator, "next", next).
+  CreateMethodProperty(realm, iterator, new StringValue(realm, 'next'), next);
+
+  // 7. Return iterator.
+  return iterator;
+}
+
+// ECMA262 7.4.8.1
+function ListIterator_next(realm: Realm): NativeFunctionValue {
+  let func = new NativeFunctionValue(realm, undefined, "next", 0, context => {
+    invariant(context instanceof ObjectValue);
+
+    // 1. Let O be the this value.
+    let O = context;
+
+    // 2. Let f be the active function object.
+    let f = func;
+
+    // 3. If O does not have a [[IteratorNext]] internal slot, throw a TypeError exception.
+    if (!O.$IteratorNext) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError, "O does not have an [[IteratorNext]] internal slot");
+    }
+
+    // 4. Let next be the value of the [[IteratorNext]] internal slot of O.
+    let next = O.$IteratorNext;
+
+    // 5. If SameValue(f, next) is false, throw a TypeError exception.
+    if (!SameValue(realm, f, next)) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError);
+    }
+
+    // 6. If O does not have an [[IteratedList]] internal slot, throw a TypeError exception.
+    if (!O.$IteratedList) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError, "O does not have an [[IteratedList]] internal slot");
+    }
+
+    // 7. Let list be the value of the [[IteratedList]] internal slot of O.
+    let list = O.$IteratedList;
+
+    // 8. Let index be the value of the [[ListIteratorNextIndex]] internal slot of O.
+    // Default to 0 for Flow.
+    let index = O.$ListIteratorNextIndex || 0;
```

Comment:
As far as I understand, you should rather state an invariant here that `typeof O.$ListIteratorNextIndex === "number"`. It cannot be `undefined`, and you shouldn't need the logic to turn that into `0`.